c2c: log roachtest throughput metrics during initial scan and cutover

msbutler opened the pull request:
This patch adds infrastructure to the cluster replication roachtest to collect
performance metrics. Specifically:
 - Initial Scan size, duration, throughput
 - Cutover size, duration, and throughput

A future patch will refactor and export these metrics to a prom instance used
to track roachtest perf over several runs. Printing these metrics is merely a
short term solution.

Informs https://github.com/cockroachdb/cockroach/issues/92697

Release note: None

msbutler committed:
c2c: log roachtest throughput metrics during initial scan and cutover

This patch adds infrastructure to the cluster replication roachtest to collect
performance metrics. Specifically:
 - Initial Scan size, duration, throughput
 - Cutover size, duration, and throughput

A future patch will refactor and export these metrics to a prom instance used
to track roachtest perf over several runs. Printing these metrics is merely a
short term solution.

Informs #92697

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94651)
<!-- Reviewable:end -->

msbutler commented:
first commit will merge in https://github.com/cockroachdb/cockroach/pull/93449

msbutler committed:
c2c: move c2c/kv0 roachtest to main c2c roachtest framework

This is nearly a pure code refactor, though it changes the roachtest to
cutover 5 minutes before the workload ends, instead of immediately after the
high water mark appears, which seems more realistic.

Release note: none

msbutler commented:
FWIW, the metrics output for the refactored c2c/kv0 looks like:
```
22:39:33 cluster_to_cluster.go:469: CutoverDuration: 0.25 Minutes
22:39:33 cluster_to_cluster.go:469: CutoverSize: 77.00 MB
22:39:33 cluster_to_cluster.go:469: CutoverThroughput: 1.70 MB/S/Node
```

lidorcarmel reviewed (commented):
do we think that disk size is good for c2c roachtests? I imagine disk writes/deletes can vary significantly because of pebble compactions. instead, can we look at c2c metrics? or addsst metrics? we should be able to know how many bytes we push.

msbutler commented on pkg/cmd/roachtest/tests/cluster_to_cluster.go:
```diff
@@ -65,8 +66,119 @@ type clusterInfo struct {
+// TODO(msbutler): move this to a general perf helper package for all roachtests
+func diskDiffThroughput(start sizeTime, end sizeTime) float64 {
+	if start.nodeCount != end.nodeCount {
+		return 0
```
I thought about this, but I'd rather report non meaningful data, instead of failing the test.

msbutler replied:
I suppose. I'll change it.

msbutler replied:
Done.

msbutler commented on pkg/cmd/roachtest/tests/cluster_to_cluster.go:
```diff
@@ -65,8 +66,119 @@ type clusterInfo struct {
+
+	populate := func(start sizeTime, end sizeTime, label string) {
+		metrics[label+"Duration"] = newMetric(end.time.Sub(start.time).Minutes(), "Minutes")
+		metrics[label+"Size"] = newMetric(float64(end.size-start.size), "MB")
```
`label+"Size"` describes the cluster size difference between two intervals. For example, it estimates the size of the initial scan. Clarified this with comment.

msbutler commented:
TFTR!!

bors r=lidorcarmel

msbutler commented:
bors retry

msbutler commented:
bors r-

msbutler committed:
c2c: fix roachtest flake

c2c/tpcc failed due to #93755. This patch fixes the roachtest.

Fixes #94877

Release note: None

Epic: None

msbutler commented:
bors retry

msbutler commented:
bors r=lidorcarmel
